### PR TITLE
[#150597257] - schedule retries when a queue action fails

### PR DESCRIPTION
**Pull request by gunzip**
Work in progress for https://www.pivotaltracker.com/story/show/150597257

We can re-schedule a message in the queue for later processing using visibilityTimeout.

Two possible retry stategies: 

1. **update** the visibilityTimeout of the received message then put it again in the queue. this strategy use the `dequeCount` property in Functions bindings and calls `context.done(err)` to return the message to the queue. after 7 days the message will be deleted from the queue in any case.

2. **create** a new message in the queue with the computed visibilityTimeout. this strategy does not rely on `dequeCount` (as it's always = 1) and calls `context.done()` to remove the old message from the queue. the number of retries is saved in the message payload.

## 1 Pros (update the same message)

- when maxDequeCount will be reached the message is moved automatically into the poison-queue
- you don't have to save the number of retries into the message payload (as dequeCount is a system property)

## 2 Pros (create a new message for every retry and delete the old one)

- you can check the number of retries looking into the message payload
- number of retries does not depened on maxDequeCount 
- you can schedule retries for more than 7 days
- context.done() does not log an error

Probably #1 is more useful if you plan to use the poison queue someway.

I chose to not use an exponential backoff strategy because the intervals between retry (in seconds) where not optimal. using a fixed number of intervals gives us more control.

see https://github.com/teamdigitale/digital-citizenship-functions/pull/133/files#diff-2d141546a695a0fa4638dedaf20abc72R9

### CODE REVIEW

File: lib/utils/azure_queues.ts [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+// Array of intervals, in seconds, to delay message processing
+// The array *must* contain = (maxDequeueCount - 1) items
+// Any value must be less than 7 days (< 604800 seconds)
+// if (MAX_RETRIES < maxDequeueCount) - 1 there will be other extraneous
+//    #(maxDequeueCount - MAX_RETRIES) retries with the default visibilityTimeout
+//    before putting the message into the poison queue
+// if (MAX_RETRIES > maxDequeueCount - 1) the system will retry even after maxDequeueCount
+//    is reached and duplicate messages will be put into the poison queue
+export const RETRIES: ReadonlyArray<number> = [
```
Comment (cloudify) [inline lib/utils/azure_queues.ts]: where do these numbers come from? truncated exp backoff is perfectly fine, see an example implementation [here](https://github.com/akka/akka/blob/master/akka-actor/src/main/scala/akka/pattern/BackoffSupervisor.scala#L149)

btw you should add a randomized factor to the delay - see page 285 of [this classic school text](http://www.uoitc.edu.iq/images/documents/informatics-institute/exam_materials/Computer%20Networks%20-%20A%20Tanenbaum%20-%205th%20edition.pdf) for an explanation

File: lib/utils/errors.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+  PermanentError = "PermanentError"
+}
+
+interface ICustomError<T extends ErrorTypes> {
+  readonly kind: T;
+  readonly message: string;
+  readonly cause?: Error;
+}
+
+const CustomError = <T extends ErrorTypes>(
```
Comment (cloudify) [inline lib/utils/errors.ts]: not sure why you need the type parameter `T`, I don't see any code that needs to be generic in `T` - it looks like the following would be enough:

``` ts
interface ICustomError {
   readonly kind: ErrorTypes;
   readonly message: string;
   readonly cause?: Error;
 }
```

Comment (gunzip) [reply lib/utils/errors.ts]: ok

File: lib/utils/azure_queues.ts (hunk 1 above)
Comment (gunzip) [reply lib/utils/azure_queues.ts]: Using Math.min(maxBackoff, Math.pow(10, retries)) was [my first try](https://github.com/gunzip/digital-citizenship-functions/blob/backoff/lib/created_message_queue_handler.ts#L104).

With this formula you get retries at [ 10, 100, 1000, 10000, 100000 ], so the last retry is at about 1.5 days from the moment the message has been enqueued. So I began tweaking the formula to get more retries, considering that we have a timespan of 7 days (max time_to_live) for each message. Finally I asked myself "how using a (tweaked) formula is better than directly specify the intervals directly ?" and went with the current implementation. 

The random delay is useful if you must avoid event collisions which is not our case IMHO: I'm not biased against adding a random factor, but note that tweaking the visibilityTimeout does not guarantee that the message will be scheduled at that exact time in the future, it only ensures that the message won't be processed *before the specified delay* (with some kind of imprecision dued to the Functions runtime). The moment the message will be dequeued depends on backpressure as well.

**Comment by gunzip** [comment]
as the [issue regarding time_to_live](https://www.pivotaltracker.com/story/show/150588117) is strictly related to this one regarding scheduled retries I was planning to solve both with one PR. At some point in time we must avoid to schedule retries that exceed the ttl.

File: lib/utils/azure_queues.ts (hunk 1 above)
Comment (cloudify) [reply lib/utils/azure_queues.ts]: untitled

Comment (cloudify) [reply lib/utils/azure_queues.ts]: `300ms * 2^count` looks quite reasonable

File: lib/utils/errors.ts (hunk 2 above)
Comment (cloudify) [reply lib/utils/errors.ts]: maybe needed in the future === not needed 😸

File: lib/utils/azure_queues.ts (hunk 1 above)
Comment (gunzip) [reply lib/utils/azure_queues.ts]: 300ms * 2^count translates into 6 retries in the first 20 seconds (aka 6 more messages to process in the queue fo each message that end up with an error). IMHO this is way too much for the kind of errors that could cause a retry (network errors, paas service downtime, etc.) which are probably restored in seconds (if not minutes or hours). moreover this will pollute the queue with a lot of messages in case of subsequent errors. 

finally, another issue with any exponential strategy is that your last retry will be at best at 3.5 days from the time the message is accepted. this means, according to our max timespan of 7 days, you'll lose 3.5 days of possible retries (for example in case we opt for 300ms^2*count the last retry is at 3.6 days, then the message will expire)

Comment (cloudify) [reply lib/utils/azure_queues.ts]: I think we are in the territory of "opinions" and unless you have some hard data to back your assumptions I would go for the [industry standard approach](https://docs.aws.amazon.com/general/latest/gr/api-retries.html) :)

Comment (cloudify) [reply lib/utils/azure_queues.ts]: does `285ms * (2 ^ count)` look better?

untitled

Comment (gunzip) [reply lib/utils/azure_queues.ts]: nicer ! BUT it looks like the visibilityTimeout granularity is in seconds (int):
http://azure.github.io/azure-storage-node/QueueService.html#updateMessage__anchor
so we must round the values or we have to chose a totally different way

Comment (cloudify) [reply lib/utils/azure_queues.ts]: just round by excess using `Math.ceil()`

Comment (gunzip) [reply lib/utils/azure_queues.ts]: I've plugged in the [backoff strategy](https://github.com/teamdigitale/digital-citizenship-functions/pull/133/files#diff-2d141546a695a0fa4638dedaf20abc72R23). 

Now we have ca. 30 retries. As maxDequeCount must be configured in host.json globally (once for every queue handler), it means that maxDequeCount must be 31, this in case you use a retry strategy OR NOT (so you'd better do : ). 

TL; DR: If you don't call retry() in your queue handlers they will retry enqueing the message 30 times in a very fast burst.

File: lib/utils/azure_queues.ts [hunk 3]
```diff
@@ -6,6 +6,12 @@ import * as winston from "winston";
 import { IContext } from "azure-functions-types";
 import { QueueService } from "azure-storage";
 
+import { unfoldr } from "fp-ts/lib/Array";
+import { none, Option, some } from "fp-ts/lib/Option";
+
+const TIME_TO_LIVE = 7 * 24 * 3600 * 1000;
+const MILLISECONDS = 285;
```
Comment (cloudify) [inline lib/utils/azure_queues.ts]: `MIN_BACKOFF_MS`

File: lib/utils/azure_queues.ts [hunk 4]
```diff
@@ -6,6 +6,12 @@ import * as winston from "winston";
 import { IContext } from "azure-functions-types";
 import { QueueService } from "azure-storage";
 
+import { unfoldr } from "fp-ts/lib/Array";
+import { none, Option, some } from "fp-ts/lib/Option";
+
+const TIME_TO_LIVE = 7 * 24 * 3600 * 1000;
```
Comment (cloudify) [inline lib/utils/azure_queues.ts]: `MAX_BACKOFF_MS`

File: lib/utils/azure_queues.ts [hunk 5]
```diff
@@ -14,16 +20,14 @@ import { QueueService } from "azure-storage";
...
-  21600,
-  86400,
-  172800,
-  345600
-];
+export const RETRIES: ReadonlyArray<number> = unfoldr((retries: number): Option<
+  [number, number]
+> => {
+  const delay = Math.ceil(MILLISECONDS * Math.pow(2, retries) / 1000);
+  return delay <= TIME_TO_LIVE
```
Comment (cloudify) [inline lib/utils/azure_queues.ts]: `TIME_TO_LIVE` is also in msec

File: lib/utils/azure_queues.ts [hunk 6]
```diff
@@ -14,16 +20,14 @@ import { QueueService } from "azure-storage";
...
-  10,
-  60,
-  600,
-  3600,
-  21600,
-  86400,
-  172800,
-  345600
-];
+export const RETRIES: ReadonlyArray<number> = unfoldr((retries: number): Option<
```
Comment (cloudify) [inline lib/utils/azure_queues.ts]: you should make this a function now `(retry: number) => number`

Comment (gunzip) [reply lib/utils/azure_queues.ts]: I need to know in advance how many retries can be scheduled in the the specified time to live (RETRIES.length) to check that maxDequeCount is setup correctly in hosts.json

Comment (cloudify) [reply lib/utils/azure_queues.ts]: `max_retries < log_2(TIME_TO_LIVE / MIN_BACKOFF)`

Comment (cloudify) [reply lib/utils/azure_queues.ts]: `max_retries = [ log_2(7 * 24 * 3600 * 1000 / 285) ] = 21`

perhaps there's a 1 off difference from your implementation, anyway it's simple algebra, you don't have to precalculate all values

Comment (gunzip) [reply lib/utils/azure_queues.ts]: of course. it was just to don't have to write the logic (pow) and the reverse logic (log) and then keep them in sync (I'll turn it to a function anyway)

File: lib/utils/azure_queues.ts [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+//    #(maxDequeueCount - MAX_RETRIES) retries with the default visibilityTimeout
+//    before putting the message into the poison queue
+// if (MAX_RETRIES > maxDequeueCount - 1) the system will retry even after maxDequeueCount
+//    is reached and duplicate messages will be put into the poison queue
+export const MAX_RETRIES = Math.floor(
+  Math.log2(MAX_BACKOFF_MS / MIN_BACKOFF_MS)
+);
+
+/**
+ * Compute the timeout in seconds before the message will be processed again
```
Comment (cloudify) [inline lib/utils/azure_queues.ts]: Add a note about the meaning of an empty option

File: lib/utils/azure_queues.ts [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+ * after an incremental delay (visibilityTimeout).
+ *
+ * Useful in case of transient errors. The message is enqueued with
+ * a newly computed visibilityTimeout (proportional to dequeueCount)
+ *
+ * @param queueService  The Azure storage queue service
+ * @param queueName     The Azure storage queue name
+ * @param context       The Functions context with bindings
+ */
+export function retry(
```
Comment (cloudify) [inline lib/utils/azure_queues.ts]: `retry` is a pretty overloaded identifier, consider perhaps a more specific identifier (e.g. `retryMessage` or a more explicit one `retryAfterDelay`)

Comment (cloudify) [inline lib/utils/azure_queues.ts]: `retry` is a pretty overloaded identifier, consider perhaps a more specific identifier (e.g. `retryMessage` or a more explicit one `retryAfterDelay`)

File: lib/utils/azure_queues.ts [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+// if (MAX_RETRIES > maxDequeueCount - 1) the system will retry even after maxDequeueCount
+//    is reached and duplicate messages will be put into the poison queue
+export const MAX_RETRIES = Math.floor(
+  Math.log2(MAX_BACKOFF_MS / MIN_BACKOFF_MS)
+);
+
+/**
+ * Compute the timeout in seconds before the message will be processed again
+ */
+export const delay = (
```
Comment (cloudify) [inline lib/utils/azure_queues.ts]: `delay` is usually used to delay an actual amount of time, perhaps `getDelayMsForRetries` will be less confusing

File: lib/utils/azure_queues.ts (hunk 7 above)
Comment (cloudify) [reply lib/utils/azure_queues.ts]: Also in the comment you're saying "seconds" whereas in the body you're using `MIN_BACKOFF_MS` without dividing by 1000 (an actual `Duration` type would be useful here) :)

File: lib/utils/azure_queues.ts [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+  context: IContext
+): void {
+  const queueMessage = context.bindingData;
+
+  winston.debug(`Retry to handle message ${queueMessageToString(context)}`);
+
+  const numberOfRetries = queueMessage.dequeueCount;
+
+  delay(numberOfRetries)
+    .map(visibilityTimeout => {
```
Comment (cloudify) [inline lib/utils/azure_queues.ts]: when you're dealing with `number` variables that represent durations it's always useful to have the unit of measure as part of the variable name, e.g. `visibilityTimeoutSec` (or `visibilityTimeoutMs`?)

File: lib/utils/errors.ts (hunk 2 above)
Comment (cloudify) [reply lib/utils/errors.ts]: I would rename `CustomError` to `RuntimeError`, seems more effective in communicating the intent
